### PR TITLE
Allow more specific entity movement customization

### LINKED ISSUE #213: Move entity options
**Is your feature request related to a problem? Please describe.**
I have a friend who wants to have the options: movePassiveMobs and moveHostileMobs. Should I do a Pull request after I wrote these options or do you think these options shouldn't be included (seperate from origin)?

Second thing I think the onlyMovePlayers CraftType option should ***not*** default to true!
File: /modules/api/.../craft/CraftTypes.java Line: 163 (position based on commit 3dd6c7f96f80b25340ac90e0b458aacddbe2801f)
`onlyMovePlayers = (boolean) data.getOrDefault("onlyMovePlayers", true);`

## Reasons
1. People could think that the moveEntity option doesn't work, because they didn't find the onlyMovePlayers option. 
```
A Person: 12.12.2019
You should put in an issue about the moveEntities property not working.    
```
2.  Normally, a ship should be able to transport peacefull creatures per default. 

**Describe the solution you'd like**
I will implement a movePassiveMobs(default: true) and moveHostileMobs(default: false) and change the default of onlyMovePlayer to false.

So every ship will be able to transport peaceful creatures, but zombies etc. won't be transported by default.

Is there a pull request interest after finishing?

Thx for reading.

### PR DESCRIPTION
**Pull request by Buh13246**
changed onlyMovePlayer to false and added a moveEntityList option which can contain multiple Class/Interface Names from package org.bukkit.entity. Default is Player and Animals.

added a new helper function in CraftType (shouldEntityBeMoved(Entity entity)  )

### Related issues:
- resolves #213 

### Checklist
- [? ] Unit tests 
- [? ] Proper internationalization
- [ X] Compiled/tested

### CODE REVIEW

File: modules/api/src/main/java/net/countercraft/movecraft/craft/CraftType.java [hunk 1]
```diff
@@ -160,7 +164,37 @@ public CraftType(File f) {
...
+            for (Object i : objList) {
+                if (!(i instanceof String)) {
+                    continue;
+                }
+                try{
+                    Class cls = cl.loadClass("org.bukkit.entity."+(String) i);
+                    moveEntityList.add(cls);
+                    System.out.println("[Craft "+f.getName()+"] moveEntityList loaded Class "+cls.getCanonicalName());
+                }catch ( ClassNotFoundException cnfe){
+                    System.out.println("ClassNotFoundException: in craft "+f.getName()+" option moveEntityList");
```
[the remark this answers is not in the document]
Comment (Buh13246) [reply modules/api/src/main/java/net/countercraft/movecraft/craft/CraftType.java]: What exact Output do you want? Error: <craft file>: invalid value <value> in moveEntityList. ?

[the remark this answers is not in the document]
Comment (Buh13246) [reply modules/api/src/main/java/net/countercraft/movecraft/craft/CraftType.java]: I don't know which Logger class you mean. Can you give me a sample line of code? Thx.

[the remark this answers is not in the document]
Comment (Buh13246) [reply modules/api/src/main/java/net/countercraft/movecraft/craft/CraftType.java]: Removed Logging completly. API can not access Movecraft.getInstance().getLogger()

**Comment by Buh13246** [comment]
@cccm5 Requested Changes made (Removed Logging completly). Sorry for delay

**Comment by Buh13246** [comment]
The 1.15 version will be easy to update. I will try to do a compatible 1.15 branch. based on this commit merged into https://github.com/eirikh1996/Movecraft/commit/9c63d91e9c2f3518f698ae7cc7b5db3affc98286

File: modules/api/src/main/java/net/countercraft/movecraft/craft/CraftType.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mwkaicz) [reply modules/api/src/main/java/net/countercraft/movecraft/craft/CraftType.java]: > 
> 
> Removed Logging completly. API can not access Movecraft.getInstance().getLogger()

How is this possible? logger must be accessible for all the time
... disabling/removing logging is bad in many ways :/

[the remark this answers is not in the document]
Comment (Buh13246) [reply modules/api/src/main/java/net/countercraft/movecraft/craft/CraftType.java]: found the default Bukkit Logger. I'm using this one now.

```Bukkit.getLogger().warning("["+f.getName()+"] moveEntityList: wrong Value: "+i);```

File: modules/api/src/main/java/net/countercraft/movecraft/craft/CraftType.java [hunk 2]
```diff
@@ -160,7 +164,37 @@ public CraftType(File f) {
...
+                    System.out.println("[Movecraft] Unknown Error on loading craft file "+f.getName());
+                    System.out.println(e.getCause().toString());
+                    System.out.println(e.getCause().getStackTrace().toString());
+                }
+            }
+
+        }else{
+            // Move Animals per default
+            moveEntityList.add( Animals.class);
+        }
```
[the remark this answers is not in the document]
Comment (Buh13246) [reply modules/api/src/main/java/net/countercraft/movecraft/craft/CraftType.java]: Removed it

**Comment by Buh13246** [comment]
Closing this now because of the lack of review. Developing now in own git repo.

**Comment by cccm5** [comment]
I apologize, I didn't see that you had updated this until now. Sorry for the lack of review. If you'd still like you can feel free to re-open this and I'll review it, however I understand if you don't want to.

**Comment by Buh13246** [comment]
okay. Sorry... I will reopen it. And I will merge it with the current Movecraft versions. I wrote it for a friends Server and they love it...

**Comment by Buh13246** [comment]
I resolved the Merge with the current version of Movecraft. Need to test it though.

**Comment by Buh13246** [comment]
@cccm5 Everything still works. Ready for merge if you don't have any more questions / improvements..

example Config:
```moveEntities: true
onlyMovePlayers: false
moveEntityList:
    - Zombie
    - Animals
    - Blaaaasdasda``` // Just to see if Errors are handled correctly

It can be every Class / Interface that is part of org.bukkit.entity
https://hub.spigotmc.org/javadocs/spigot/org/bukkit/entity/package-summary.html

Animals will move all subinterfaces of Animals (
    AbstractHorse, Bee, Cat, ChestedHorse, Chicken, Cow, Donkey, Fox, Horse, Llama, Mule, MushroomCow, Ocelot, Panda, Parrot, Pig, PolarBear, Rabbit, Sheep, SkeletonHorse, Tameable, TraderLlama, Turtle, Wolf, ZombieHorse)
Zombie will move all Zombie Interfaces (Drowned, Husk, PigZombie, ZombieVillager)

The interfaces are dynamicly loaded and every LivingEntity in new versions work exactly the same than old Interfaces, without modifying any Code...

**Comment by Buh13246** [comment]
@cccm5 You said that the default should be only Players, but I think the default should be all Entities (Backwards Compability). If you set onlyMovePlayers=false and moveEntities=true, than the craft should move all Enitys be default, right?

**Comment by Buh13246** [comment]
> could we instead use an enum set

I think than we would lost the possability to say all Animals or all Spiders, right?

**Comment by Buh13246** [comment]
> > could we instead use an enum set
> 
> I think than we would lost the possability to say all Animals or all Spiders, right?
> But performance is definitly a big thing to worry about, so I will look into it...
What do you think about building a Set of EntityTypes that would be afected be moving. Than we have a performance Overhead on loading, but the check if an Entity should be Moved is just a contains of the Set.

**Comment by cccm5** [comment]
Ya that's what I was getting at, we should be building EnumSet's based on a hierarchy of the backing class types (i.e. generate from `Spiders` we create a set that contains all the EntityTypes of the various spiders), and than store the unions of these sets in the CraftType. At the moment you have both a linear lookup and an expensive class accessor, which will probably add a noticeable performance overhead to teleporting.

**Pull request closed without merging**